examples/OpticalFlow/flownet_models.py

```diff
@@ -0,0 +1,420 @@
+
+    def get(y, x):
+        idx = tf.reshape(batch_ids * h * w + y * w + x, [-1])
+        idx = tf.cast(idx, tf.int32)
+        return tf.gather(img_flat, idx)
+
+    val = tf.zeros_like(alpha)
+    val += (1 - alpha) * (1 - beta) * tf.reshape(get(yT, xL), [-1, h, w, c])
+    val += (0 + alpha) * (1 - beta) * tf.reshape(get(yT, xR), [-1, h, w, c])
+    val += (1 - alpha) * (0 + beta) * tf.reshape(get(yB, xL), [-1, h, w, c])
+    val += (0 + alpha) * (0 + beta) * tf.reshape(get(yB, xR), [-1, h, w, c])
+
+    # we need to enforce the channel_dim known during compile-time here
+    shp = img.shape.as_list()
+    return tf.reshape(tf.transpose(val, [0, 3, 1, 2]), [-1, shp[1], h, w])
+
+
+def resize(x, factor=4, mode='bilinear'):
+    """Resize input tensor with unkown input-shape by a factor
+
+    Args:
+        x (tf.Tensor): tensor NCHW
+        factor (int, optional): resize factor for H, W
+
+    Note:
+        Differences here against Caffe have huge impacts on the
+        quality of the predictions.
+
+    Returns:
+        tf.Tensor: resized tensor NCHW
+    """
+    assert mode in ['bilinear', 'nearest']
+    shp = tf.shape(x)[2:] * factor
+    # NCHW -> NHWC
+    x = tf.transpose(x, [0, 2, 3, 1])
+    if mode == 'bilinear':
+        x = tf.image.resize_bilinear(x, shp, align_corners=True)
+    else:
+        # better approximation of what Caffe is doing
+        x = tf.image.resize_nearest_neighbor(x, shp, align_corners=False)
+    # NHWC -> NCHW
+    return tf.transpose(x, [0, 3, 1, 2])
+
+
+class FlowNetBase(ModelDesc):
+    def __init__(self, height=None, width=None, channels=3):
+        self.height = height
+        self.width = width
+        self.channels = channels
+
+    def inputs(self):
+        return [tf.placeholder(tf.float32, (1, self.channels, self.height, self.width), 'left'),
+                tf.placeholder(tf.float32, (1, self.channels, self.height, self.width), 'right')]
+
+    def graph_structure(self, x):
+        raise NotImplementedError()
+
+    def build_graph(self, left, right):
+        x = tf.stack([left, right], axis=2)
+        rgb_mean = tf.reduce_mean(x, axis=[2, 3, 4], keep_dims=True)
+        x = (x - rgb_mean) / 255.
+
+        prediction = self.graph_structure(x)
+        prediction = resize(prediction / 20.)
+        tf.identity(prediction, name="prediction")
+
+
+class FlowNet2(FlowNetBase):
+
+    def build_graph(self, left, right):
+        x = tf.stack([left, right], axis=2)
+        rgb_mean = tf.reduce_mean(x, axis=[2, 3, 4], keep_dims=True)
+        x = (x - rgb_mean) / 255.
+
+        prediction = self.graph_structure(x)
+        tf.identity(prediction, name="prediction")
+
+    def graph_structure(self, x):
+        x1, x2 = tf.unstack(x, axis=2)
+        x1x2 = tf.concat([x1, x2], axis=1)
+
+        # FlowNet-C
+        flownetc_flow2 = FlowNet2C().graph_structure(x)
+        flownetc_flow = resize(flownetc_flow2 * 20., mode='bilinear')
```

Comment:
Could you make "20" a constant variable, and add some words and references about it? That would make me more comfortable seeing it around.
Also, is it redundant here to multiple it by 20 and then divide by 20 later?